### PR TITLE
[Force Push Candidate] Asynchronous Tooling for Policies

### PR DESCRIPTION
**Pull request by KPrasch**

### CODE REVIEW

File: nucypher/policy/policies.py [hunk 1]
```diff
@@ -391,7 +415,8 @@ def enact(self, network_middleware, publish_treasure_map=True) -> dict:
             self.alice.add_active_policy(self)
 
             if publish_treasure_map is True:
-                return self.publish_treasure_map(network_middleware=network_middleware)
+                publication_result = self.inject_alice_publication_threadpool_into(self.publish_treasure_map, network_middleware=network_middleware)
+                # return self.publish_treasure_map(network_middleware=network_middleware)
```
Comment (derekpierre) [inline nucypher/policy/policies.py:419]: we no longer return anything here?

File: nucypher/policy/policies.py [hunk 2]
```diff
@@ -165,30 +165,100 @@ def payload(self):
...
+        if self.total_disposed() < len(self.nodes):
+            _ = self._completion_queue.get()  # Interesting opportuntiy to pass some data, like the list of contacted nodes above.
+
+    def _handle_success(self, response, node):
+        if response.status_code == 202:
+            self.completed[node] = response
+        else:
+            assert False  # TODO: What happens if this is a 300 or 400 level response?
+
+        if len(self.completed) == self._block_until_this_many_are_complete:
```
Comment (derekpierre) [inline nucypher/policy/policies.py:229]: Because of concurrency should this be the following for safety...
```suggestion
        if len(self.completed) >= self._block_until_this_many_are_complete:
```

File: nucypher/policy/policies.py [hunk 3]
```diff
@@ -165,30 +165,100 @@ def payload(self):
...
+
+    def _engage_node(self, node):
+        maybe_coro = self.f(node, network_middleware=self.network_middleware, *self.args, **self.kwargs)
+
+        d = ensureDeferred(maybe_coro)
+        d.addCallback(self._handle_success, node)
+        d.addErrback(self._handle_error, node)
+        return d
+
+    def start(self):
```
Comment (derekpierre) [inline nucypher/policy/policies.py:258]: Perhaps an associated `stop()` function could be provided that unblocks everything , stops the threads and threadpool could be good for forcibly stopping the mutex in extreme cases...particularly in failed tests?

File: docs/source/guides/network_node/staking_guide.rst [hunk 4]
```diff
@@ -79,6 +79,8 @@ All staking-related operations done by Staker are performed through the ``nucyph
 +----------------------+-------------------------------------------------------------------------------+
 |  ``mint``            | Mint last portion of reward                                                   |
 +----------------------+-------------------------------------------------------------------------------+
+|  ``merge``           | Merge two stakes into one                                                     |
```
Comment (KPrasch) [inline docs/source/guides/network_node/staking_guide.rst:82]: Hm - I dont expect to see this diff here - perhaps check the base and history again.

File: examples/run_demo_ursula_fleet.py [hunk 5]
```diff
@@ -17,15 +17,15 @@
...
 from nucypher.characters.lawful import Ursula
+from nucypher.config.constants import APP_DIR
 
 FLEET_POPULATION = 12
 DEMO_NODE_STARTING_PORT = 11500
 
 ursula_maker = partial(Ursula, rest_host='127.0.0.1',
                        federated_only=True,
-                       domains=[TEMPORARY_DOMAIN]
+                       domains=[":TEMPORARY_DOMAIN:"]
```
Comment (KPrasch) [inline examples/run_demo_ursula_fleet.py:28]: Why not use the constant?

File: nucypher/blockchain/eth/actors.py [hunk 6]
```diff
@@ -876,19 +878,38 @@ def current_stake(self) -> NU:
         """The total number of staked tokens, i.e., tokens locked in the current period."""
         return self.locked_tokens(periods=0)
 
-    def stakes_filtered_by_status(self, parent_status: Stake.Status) -> Iterable[Stake]:
-        """Returns stakes for this staker which have specified or child status."""
+    def filtered_stakes(self,
```
Comment (KPrasch) [inline nucypher/blockchain/eth/actors.py:881]: Again - I do not expect to see this addition ( not part of this PR )

File: nucypher/policy/policies.py (hunk 1 above)
Comment (jMyles) [reply nucypher/policy/policies.py]: Yeah, I've gone back and forth.  What do you think?  At this moment, this function works for my mental model as a void.  If the middleware starts returning a deferred (toward which I'm increasingly leaning), then we will obviously reevaluate.

File: nucypher/policy/policies.py (hunk 2 above)
Comment (jMyles) [reply nucypher/policy/policies.py:245]: Hmm, no that's no good either, as it will put additional values through the queue, emit duplicate log emissions, etc.  We need to use the `_released` flag here.  Wanna open an issue?

File: nucypher/policy/policies.py (hunk 3 above)
Comment (jMyles) [reply nucypher/policy/policies.py:281]: :+1: 

Care to open an issue?

File: examples/run_demo_ursula_fleet.py [hunk 7]
```diff
@@ -36,18 +36,15 @@ def spin_up_federated_ursulas(quantity: int = FLEET_POPULATION):
 
     ursulas = []
 
-    sage = ursula_maker(
-        rest_port=ports[0],
-        db_filepath="sage.db",
-    )
+    sage = ursula_maker(rest_port=ports[0], db_filepath=f"{APP_DIR.user_cache_dir}/sage.db")
```
[the remark this answers is not in the document]
Comment (jMyles) [reply examples/run_demo_ursula_fleet.py:39]: Ahh, yes, I initially added this locally (thinking about @vzotova :-) ), and then I guess I inadvertently removed it when I changed from the local dir to `APP_DIR`.

File: examples/run_demo_ursula_fleet.py [hunk 8]
```diff
@@ -36,18 +36,15 @@ def spin_up_federated_ursulas(quantity: int = FLEET_POPULATION):
...
 
     ursulas.append(sage)
     for index, port in enumerate(ports[1:]):
         u = ursula_maker(
             rest_port=port,
             seed_nodes=[sage.seed_node_metadata()],
             start_learning_now=True,
-            db_filepath=f"{port}.db",
+            db_filepath=f"{APP_DIR.user_cache_dir}/{port}.db",
```
[the remark this answers is not in the document]
Comment (jMyles) [reply examples/run_demo_ursula_fleet.py:47]: Oh, right.

File: nucypher/characters/lawful.py [hunk 9]
```diff
@@ -131,8 +128,10 @@ def __init__(self,
             self.m = m
             self.n = n
 
-            self.publication_threadpool = ThreadPool(maxthreads=120, name="Alice Policy Publication")  # In the future, this value is perhaps best set to something like 3-4 times the optimal "high n", whatever we determine that to be.
-            self.publication_threadpool.start()
+            self._policy_queue = Queue()
+            self._policy_queue.put(READY)
+            # self.publication_threadpool = ThreadPool(maxthreads=120, name="Alice Policy Publication")  # In the future, this value is perhaps best set to something like 3-4 times the optimal "high n", whatever we determine that to be.
+            # self.publication_threadpool.start()
```
[the remark this answers is not in the document]
Comment (jMyles) [reply nucypher/characters/lawful.py:134]: Yep.

File: nucypher/characters/lawful.py [hunk 10]
```diff
@@ -275,6 +287,7 @@ def grant(self,
               discover_on_this_thread: bool = True,
               timeout: int = None,
               publish_treasure_map: bool = True,
+              block_for_a_little_while: bool = True,
```
[the remark this answers is not in the document]
Comment (jMyles) [reply nucypher/characters/lawful.py:290]: Fine.  But first you have to watch this: https://www.youtube.com/watch?v=OkSLswPSq2o

Also, it really is just totally arbitrary, so I'm not sure anything like "min publication" even really works.

File: docs/source/guides/network_node/staking_guide.rst (hunk 4 above)
Comment (jMyles) [reply docs/source/guides/network_node/staking_guide.rst:82]: It's rebased over `main`, while the current target (TI) hasn't been.

File: examples/run_demo_ursula_fleet.py (hunk 5 above)
Comment (jMyles) [reply examples/run_demo_ursula_fleet.py:28]: Yeah... I was getting an error when trying to do that when it casts the constant to bytes because it tries to call its `encode` method.  (Which, actually, might be something we want to consider adding in CS).  Change it back and see for yourself.  I haven't tracked down how it started.

File: nucypher/policy/policies.py (hunk 3 above)
Comment (derekpierre) [reply nucypher/policy/policies.py:281]: Filed #2189

File: nucypher/policy/policies.py (hunk 2 above)
Comment (derekpierre) [reply nucypher/policy/policies.py:245]: Filed https://github.com/nucypher/nucypher/issues/2190.

File: nucypher/policy/policies.py [hunk 11]
```diff
@@ -303,17 +406,22 @@ def publish_treasure_map(self, network_middleware: RestMiddleware, blockchain_si
...
 
-        publication_deferreds = list()
-        for node in self.bob.matching_nodes_among(self.alice.known_nodes):
+        target_nodes = self.bob.matching_nodes_among(self.alice.known_nodes)
+        self.publishing_mutex = NodeEngagementMutex(f=self.put_treasure_map_on_node,
+                                                    nodes=target_nodes,
+                                                    network_middleware=network_middleware,
+                                                    percent_to_complete_before_release=10)
+
+        self.publishing_mutex.start()
```
Comment (derekpierre) [inline nucypher/policy/policies.py:415]: This method purports to return a `NodeEngagementMutex` but we don't return it. Is that still the plan or should the method return `None`?

File: nucypher/policy/policies.py (hunk 1 above)
Comment (derekpierre) [reply nucypher/policy/policies.py]: My worry stemmed from the fact that this method (`enact()`) is marked to return a `dict`. The void is fine, but does the signature need to be changed?

File: nucypher/characters/lawful.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/characters/lawful.py:290]: Haha 👀 - vague indeed!

Perhaps `completion/engagement` instead of `publication`

File: nucypher/policy/policies.py (hunk 1 above)
Comment (jMyles) [reply nucypher/policy/policies.py]: If this change is accepted, then yes.  (There's lots of that variety of cleanup that will be required on TI; I haven't run mypy on things yet.)

File: nucypher/characters/lawful.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jMyles) [reply nucypher/characters/lawful.py:290]: `block_until_min_engagement` ?  What's the minimum engagement?  And why?

It's really more like, `block_until_success_is_reasonably_likely`

[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/characters/lawful.py:290]: 👌

File: nucypher/network/nodes.py [hunk 12]
```diff
@@ -903,30 +872,31 @@ def learn_from_teacher_node(self, eager=False, canceller=None):
...
+                self.log.warn(f'Verification Failed - '
+                              f'{sprout} has an invalid wallet signature for {sprout.decentralized_identity_evidence}')
+
+            except sprout.UnbondedWorker:
+                self.log.warn(f'Verification Failed - '
+                              f'{sprout} is not bonded to a Staker.')
+
+            # TODO: Handle invalid sprouts
+            # except sprout.Invalidsprout:
+            #     self.log.warn(sprout.invalid_metadata_message.format(sprout))
```
Comment (cygnusv) [inline nucypher/network/nodes.py:894]: Is this TODO tracked somewhere? If so, add the issue number.

File: nucypher/network/nodes.py [hunk 13]
```diff
@@ -849,22 +849,19 @@ def learn_from_teacher_node(self, eager=False, canceller=None):
...
 
         try:
             self.verify_from(current_teacher, node_payload, signature=signature)
-        except InvalidSignature as e:
-            self.suspicious_activities_witnessed['vladimirs'].append(
-                ('Node payload improperly signed', node_payload, signature))
-            self.log.warn(
-                f"Invalid signature ({signature}) received from teacher {current_teacher} for payload {node_payload}")
+        except Learner.InvalidSignature:  # TODO: Ensure wev've got the right InvalidSignature exception here
+            self.suspicious_activities_witnessed['vladimirs'].append(('Node payload improperly signed', node_payload, signature))
```
Comment (cygnusv) [inline nucypher/network/nodes.py]: We have:
* `Character.verify_from()` that raises `cryptography.exceptions.InvalidSignature`
* `Learner.verify_from()` that raises `Learner.InvalidSignature`
In the context of an object that is both a Character and a Learner, it seems extremely confusing.

File: tests/acceptance/network/test_network_actors.py [hunk 14]
```diff
@@ -162,6 +162,7 @@ class FakeArrangement:
                                                    ursula=vladimir)
 
 
+@pytest.mark.skip('Needs to be restored -- no way to access treasure maps from stranger ursulas here.')
```
Comment (cygnusv) [inline tests/acceptance/network/test_network_actors.py:165]: Needs issue tracking

File: nucypher/characters/lawful.py [hunk 15]
```diff
@@ -14,25 +14,24 @@
 You should have received a copy of the GNU Affero General Public License
 along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
 """
-
+import contextlib
 import json
 import random
 from base64 import b64encode, b64decode
 from base64 import b64decode, b64encode
```
Comment (cygnusv) [inline nucypher/characters/lawful.py:21]: Duplicated imports
```suggestion
```

File: nucypher/policy/policies.py [hunk 16]
```diff
@@ -316,6 +316,14 @@ def publish_treasure_map(self, network_middleware: RestMiddleware, blockchain_si
         self.publishing_mutex = PolicyPayloadMutex(publication_deferreds, percent_to_complete_before_release=10)
         # return self.publishing_mutex  # I dunno.. return this?  Why not just use the composed version?
 
+    def inject_alice_publication_threadpool_into(self, f, *args, **kwargs):
+        d = ensureDeferred(self.alice.get_publication_threadpool(policy_id=self.id.hex()[0:10]))
+        d.addCallback(f, *args, **kwargs)
+        def huh(result):
+            assert False
```
Comment (cygnusv) [inline nucypher/policy/policies.py:422]: Maybe something different than an AssertionError?

File: examples/run_demo_ursula_fleet.py (hunk 5 above)
Comment (cygnusv) [reply examples/run_demo_ursula_fleet.py:28]: Maybe that's a reason for using an actual constant, and not a CS constant.

**Review by cygnusv — APPROVED** [review]
Wonderful, and even fun to read sometimes!

File: nucypher/network/nodes.py (hunk 13 above)
Comment (jMyles) [reply nucypher/network/nodes.py]: Yes, please open an issue for this.

**Comment by jMyles** [comment]
Force-pushed over TI.  See `aaa2d817f`
